src/v/cluster/producer_state.cc

```diff
@@ -0,0 +1,264 @@
+
+#include "producer_state.h"
+
+#include "cluster/logger.h"
+#include "cluster/producer_state_manager.h"
+#include "vassert.h"
+
+namespace cluster {
+
+result_promise_t::future_type request::result() const {
+    return _result.get_shared_future();
+}
+
+std::optional<request_ptr> requests::last_request() const {
+    if (!_inflight_requests.empty()) {
+        return _inflight_requests.back();
+    } else if (!_finished_requests.empty()) {
+        return _finished_requests.back();
+    }
+    return std::nullopt;
+}
+
+bool requests::is_valid_sequence(seq_t incoming) const {
+    auto last_req = last_request();
+    return
+      // this is the first request with seq=0
+      (!last_req && incoming == 0)
+      // incoming request forms a sequence with last_request
+      || (last_req && last_req.value()->_last_sequence + 1 == incoming)
+      // sequence numbers got rolled over because they hit int32 max limit.
+      || (last_req && last_req.value()->_last_sequence == std::numeric_limits<seq_t>::max() && incoming == 0);
+}
+
+result<request_ptr> requests::try_emplace(
+  seq_t first, seq_t last, model::term_id current, bool reset) {
+    if (reset) {
+        // reset all the sequence tracking state, avoids any sequence
+        // checks for sequence tracking.
+        while (!_inflight_requests.empty()) {
+            if (!_inflight_requests.front()->has_completed()) {
+                _inflight_requests.front()->set_value(errc::timeout);
+            }
+            _inflight_requests.pop_front();
+        }
+        _finished_requests.clear();
+    } else {
+        // gc and fail any inflight requests from old terms
+        // these are guaranteed to be failed because of sync() guarantees
+        // prior to this request.
+        while (!_inflight_requests.empty()
+               && _inflight_requests.front()->_term < current) {
+            if (!_inflight_requests.front()->has_completed()) {
+                // Here we know for sure the term change, these in flight
+                // requests are going to fail anyway, mark them so.
+                _inflight_requests.front()->set_value(errc::timeout);
+            }
+            _inflight_requests.pop_front();
+        }
+
+        // check if an existing request matches
+        auto match_it = std::find_if(
+          _finished_requests.begin(),
+          _finished_requests.end(),
+          [first, last](const auto& request) {
+              return request->_first_sequence == first
+                     && request->_last_sequence == last;
+          });
+
+        if (match_it != _finished_requests.end()) {
+            return *match_it;
+        }
+
+        match_it = std::find_if(
+          _inflight_requests.begin(),
+          _inflight_requests.end(),
+          [first, last, current](const auto& request) {
+              return request->_first_sequence == first
+                     && request->_last_sequence == last
+                     && request->_term == current;
```

Comment:
we already wiped all old terms so `request->_term == current` condition should be always true if so it's better to remove it from the filter but to add an assertion once there is a match